### PR TITLE
Per request id metrics for VRF/Runlog tests

### PR DESCRIPTION
**Pull request by skudasov**

### CODE REVIEW

File: suite/performance/common.go [hunk 1]
```diff
@@ -99,3 +108,246 @@ func recordResults(b ginkgo.Benchmarker, ID string, results []time.Duration) err
...
+}
+
+// GetJobID gets internal job id
+func (n RunlogNodeData) GetJobID() string {
+	return n.JobID
+}
+
+// GetProvingKeyHash gets proving key hash for VRF
+func (n RunlogNodeData) GetProvingKeyHash() [32]byte {
+	return [32]byte{}
```
Comment (kalverra) [inline suite/performance/common.go:288]: Why is this empty? Does runlog not actually have one of these?

File: suite/performance/common.go [hunk 2]
```diff
@@ -99,3 +108,246 @@ func recordResults(b ginkgo.Benchmarker, ID string, results []time.Duration) err
...
+			for k, v := range m {
+				c[contractAddr][k] = v
+			}
+		}
+	}
+}
+
+// LimitErrGroup implements the errgroup.Group interface, but limits goroutines to a
+// execute at a max throughput.
+type LimitErrGroup struct {
```
Comment (kalverra) [inline suite/performance/common.go:326]: This is more for my understanding: Why rate limit the errors?

File: suite/performance/runlog.go [hunk 3]
```diff
@@ -71,14 +67,22 @@ func NewRunlogTest(
...
 		Deployer:    deployer,
 		adapter:     adapter,
-		jobMap:      RunlogJobMap{},
+		testResults: NewPerfRequestIDTestResults(),
+		jobMap:      ContractsNodesJobsMap{},
 	}
 }
 
-// RecordValues records Runlog metrics
+// RecordValues will query all of the latencies of the VRFConsumer and match them by RequestID
```
Comment (kalverra) [inline suite/performance/runlog.go:75]: > will query all of the latencies of the `VRFConsumer`

I'm assuming typo, this looks pretty similar to how VRF operates though, right?

File: suite/performance/runlog.go [hunk 4]
```diff
@@ -171,27 +174,74 @@ func (f *RunlogTest) Run() error {
...
 	for {
 		select {
 		case <-ctx.Done():
 			log.Warn().Msg("Test finished")
+			time.Sleep(f.TestOptions.GracefulStopDuration)
+			cancelPerfEvents()
 			return nil
 		default:
-			log.Warn().Int("RoundID", i).Msg("New round")
+			log.Warn().Int("RoundID", currentRound).Msg("New round")
```
Comment (kalverra) [inline suite/performance/runlog.go:195]: Why does this necessitate a `WARN` log?

File: suite/performance/runlog.go [hunk 5]
```diff
@@ -171,27 +174,74 @@ func (f *RunlogTest) Run() error {
...
 			if err := f.requestData(); err != nil {
 				return err
 			}
-			if err := f.waitRoundEnd(i); err != nil {
+			if err := f.waitRoundEnd(currentRound + 1); err != nil {
 				return err
 			}
-			i++
+			if f.TestOptions.NumberOfRounds != 0 && currentRound >= f.TestOptions.NumberOfRounds {
+				log.Warn().Msg("Final round is reached")
```
Comment (kalverra) [inline suite/performance/runlog.go:203]: Same here, this seems like a good thing?

**Review by kalverra — COMMENTED** [review]
Left a few comments, need to spend a bit more time with the code and you to grasp some more of the big picture stuff.

File: suite/performance/runlog.go (hunk 4 above)
Comment (skudasov) [reply suite/performance/runlog.go:195]: Just visually better, as a marker for long tests, can switch to INFO

File: suite/performance/common.go (hunk 2 above)
Comment (skudasov) [reply suite/performance/common.go:326]: Just a generic goroutine group with a rate-limiting so we don't hit node hard

File: suite/performance/common.go (hunk 1 above)
Comment (skudasov) [reply suite/performance/common.go:288]: Nope, kinda ugly thing, but I hope sometimes we can separate metrics collecting from test logic/deployment completely

Comment (skudasov) [reply suite/performance/common.go:288]: Will fix a comment though

**Review by kalverra — APPROVED** [review]

**Review by tateexon — APPROVED** [review]

**Pull request merged**